Python code execution tool (System/Venv/Docker)

marklysze opened the pull request:
## Why are these changes needed?

Tool for code execution that utilises an environment object to determine where execution takes place. This can be the local python environment (aka System), a virtual environment (aka Venv), or a Docker container.

This will replace the need for the code execution capabilities currently in ConversableAgent. They will be useful for building agents for things like code development, code testing, etc.

## Related issue number

N/A

## Checks

- [ ] I've included any doc changes needed for https://docs.ag2.ai/. See https://docs.ag2.ai/docs/contributor-guide/documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

marklysze commented:
Sample code:
```
from autogen import ConversableAgent, register_function
from autogen.tools.experimental import PythonLocalExecutionTool

# Initialize the tool
python_executor = PythonLocalExecutionTool(
    use_venv=True,
    venv_path="/app/ag2/ms_code/.venv_ms_testing",  # Use an existing virtual environment (it will create one on each call otherwise, destroying it upon completion)
    timeout=60,
)

llm_config = {"model": "gpt-4o-mini", "api_type": "openai"}

# Create an agent that can use the tool
code_runner = ConversableAgent(
    name="code_runner",
    system_message="You are a code executor agent, when you don't execute code write the message 'TERMINATE' by itself.",
    llm_config=llm_config,
)

question_agent = ConversableAgent(
    name="question_agent",
    system_message="You are a developer AI agent. Send all your code suggestions to the python_executor tool where it will be executed and result returned to you. Keep refining the code until it works.",
    llm_config=llm_config,
)

register_function(
    python_executor,
    caller=question_agent,
    executor=code_runner,
    description="Run Python code",
)

result = code_runner.initiate_chat(
    recipient=question_agent,
    message="""
    Write a Python code with incorrect syntax. Always install numpy and pandas.
    """,
    max_turns=5,
)
```

davorrunje commented:
We can use [`asyncer`](https://asyncer.tiangolo.com/tutorial/asyncify/#use-asyncify) to support async file handling and remove the optional dependency to `aiofiles`. E.g.

```python
from asyncer import asyncify

def read_file(name: str) -> str:
    with open(name) as f:
        return f.read()


async def a_read_file(name: str):
    content = await asyncify(read_file)(name="name")
    print(content)

```

marklysze commented:
> We can use [`asyncer`](https://asyncer.tiangolo.com/tutorial/asyncify/#use-asyncify) to support async file handling and remove the optional dependency to `aiofiles`. E.g.
> 
> ```python
> from asyncer import asyncify
> 
> def read_file(name: str) -> str:
>     with open(name) as f:
>         return f.read()
> 
> 
> async def a_read_file(name: str):
>     content = await asyncify(read_file)(name=name)
>     print(content)
> 
> ```

Great, thanks!

davorrunje commented:
I think we should use context managers to define things like python environment and working directories. They need to have lifecycle management and that should be made explicit, not hidden in the implementational details inside `ConversableAgent`.

This is how it could look like:

``` python
from autogen import ConversableAgent, LLMConfig
from autogen.environments import PythonEnvironment, WorkingDirectory
from autogen.tools.experimental import PythonLocalExecutionTool

# create virtual env and manage its lifecycle
with PythonEnvironment(python_version="3.11", dependancies=["numpy>2.1,<3", "pandas"]):

    # create working directories
    with WorkingDirectory.create_tmp() as wd1, WorkingDirectory.create_tmp() as wd2:
        # Initialize the tool
        python_executor = PythonLocalExecutionTool(
            use_venv=True,
            timeout=60,
            # the default working directory is from the outer scope (wd2), but you can change it explicitely
            working_directory=wd1,
            # it is using the python environment from the outer scope, but we could change it
           # python_environment=...
        )

        with LLMConfig(model="gpt-4o-mini", api_type= "openai"):

            # Create an agent that can use the tool
            code_runner = ConversableAgent(
                name="code_runner",
                system_message="You are a code executor agent, when you don't execute code write the message 'TERMINATE' by itself.",
                tools=python_executor,
            )

            question_agent = ConversableAgent(
                name="question_agent",
                system_message="You are a developer AI agent. Send all your code suggestions to the python_executor tool where it will be executed and result returned to you. Keep refining the code until it works.",
            )

        result = code_runner.initiate_chat(
            recipient=question_agent,
            message="""
            Write a Python code with incorrect syntax. Always install numpy and pandas.
            """,
            max_turns=5,
        )
```

marklysze commented:
@davorrunje I've updated, incorporated environments. If you can have a look over please. Also missing are tests, would like some guidance on how to test this.

marklysze commented:
DockerPythonEnvironment example:
```
from autogen import ConversableAgent, LLMConfig, register_function

# Import the environment, working directory, and code execution tool
from autogen.environments import DockerPythonEnvironment, WorkingDirectory
from autogen.tools.experimental import PythonCodeExecutionTool

with DockerPythonEnvironment(image="python:3.11-slim", pip_packages=["numpy", "pandas", "matplotlib"]) as docker_env:
    with WorkingDirectory(path="/tmp/ag2_working_dir/") as wd:
        # Create our code execution tool, using the environment and working directory from the above context managers
        python_executor = PythonCodeExecutionTool(
            timeout=60,
            # If not using the context managers above, you can set the working directory and python environment here
            # working_directory=wd,
            # python_environment=docker_env,
        )

    with LLMConfig(model="gpt-4o", api_type="openai"):
        # code_runner has the code execution tool available to execute
        code_runner = ConversableAgent(
            name="code_runner",
            system_message="You are a code executor agent, when you don't execute code write the message 'TERMINATE' by itself.",
            human_input_mode="NEVER",
        )

        # question_agent has the code execution tool available to its LLM
        question_agent = ConversableAgent(
            name="question_agent",
            system_message=(
                "You are a developer AI agent. "
                "Send all your code suggestions to the python_executor tool where it will be executed and result returned to you. "
                "Keep refining the code until it works."
            ),
        )

    # Register the python execution tool with the agents
    register_function(
        python_executor,
        caller=question_agent,
        executor=code_runner,
        description="Run Python code",
    )

    result = code_runner.initiate_chat(
        recipient=question_agent,
        message=(
            "Write Python code to print the current Python version followed by the numbers 1 to 11. "
            "Make a syntax error in the first version and fix it in the second version."
        ),
        max_turns=5,
    )

    print(f"Result: {result.summary}")
    ```

marklysze commented:
Note: I want to refactor the Docker configuration parameters into two new classes, DockerExistingContainerConfig and DockerNewContainerConfig, as the `DockerPythonEnvironment` parameter list is long.

radman-x commented:
Would be nice if there was an async version that kept the executing code running so, eg, a console python script could be started and the coder agent could repeatedly send in user input, get results, analyze those results, send in more input, etc. while the subprocess is still running. This would be awesome for unit/functional testing. May be missing something obvious. Briefly looked through code and couldn’t see how to do this. If this is already supported, can you give a small sample code how to set up?

qingyun-wu commented:
Nice one! Hopefully, we can find a chance to continue this soon!

qingyun-wu commented:
@sonichi when you have a moment, could you help review this too? Thanks!
